Support dispatching blocked jobs directly from the UI

lexcao opened the pull request:
Introduce the dispatch action to blocked jobs

Closes #37

![image](https://github.com/basecamp/mission_control-jobs/assets/14901139/9804aaf0-8c6c-4505-b3b7-dd39e2cf8d91)

lexcao commented:
Hi @rosa 
I am not sure the suitable way to `dispatch` a job directly, correct me if I am wrong.
```
def dispatch_job(job, jobs_relation)
  job = find_solid_queue_job!(job.job_id, jobs_relation)
  job.dispatch_bypassing_concurrency_limits
  job.blocked_execution.destroy!
end
```

And, can you guide me on writing some tests about that?

rosa commented:
Hey @lexcao, that'd work, but you'd need a transaction to ensure you don't dispatch the job leaving the blocked execution behind. You'd also want a more descriptive name for the method, something like `dispatch_blocked_job_immediately`. 

To write tests for that, you could perhaps [check the tests for concurrency controls in Solid Queue](https://github.com/basecamp/solid_queue/blob/01b3a5d57053093d1459171013c5417b7252c4a6/test/integration/concurrency_controls_test.rb).

lexcao commented:
Hi @rosa 
Thanks for guiding, I have added a simple test.
And the PR is ready for review.

rosa commented on lib/active_job/queue_adapters/solid_queue_ext.rb:
```diff
@@ -82,6 +82,12 @@ def discard_job(job, jobs_relation)
+  def dispatch_job(job, jobs_relation)
+    job = find_solid_queue_job!(job.job_id, jobs_relation)
+
+    dispatch_blocked_job_immediately(job) if job.blocked?
```
Do you need the `job.blocked?` check here if the `jobs_relation` that you're receiving is already for blocked jobs? That's it, why couldn't this be the same as this method, where we don't check `job.failed?` before calling `retry`: 
```ruby
  def retry_job(job, jobs_relation)
    find_solid_queue_job!(job.job_id, jobs_relation).retry
  end
```

rosa commented on lib/active_job/queue_adapters/resque_ext.rb:
```diff
@@ -77,6 +77,10 @@ def discard_job(job, jobs_relation)
 
+  def dispatch_job(job, jobs_relation)
+    # noop
+  end
```
I think we don't want this here because we already have a way to express that a method is not supported by an adapter and does nothing, via the `Adapter` module. In the `Adapter` module, you get methods such as 
```ruby
  def workers
    if exposes_workers?
      raise_incompatible_adapter_error_from :workers
    end
  end
```
that, in the case of Resque right now, it just returns `nil` because `exposes_workers?` is `false` for resque. We want something similar in this case. If the adapter supports the `blocked` status, then it should provide this method. If it doesn't support it, we don't need to add it and no-op.

rosa commented on test/active_job/queue_adapters/adapter_testing/dispatch_jobs.rb:
```diff
@@ -0,0 +1,15 @@
+module ActiveJob::QueueAdapters::AdapterTesting::DispatchJobs
+  extend ActiveSupport::Concern
```
Oh, I know you copied this one from other similar modules to test the adapter, but I believe we don't need this, since we aren't using any of the `ActiveSupport::Concern` features.

rosa commented on test/active_job/queue_adapters/adapter_testing/dispatch_jobs.rb:
```diff
@@ -0,0 +1,15 @@
+
+    blocked_jobs.each(&:dispatch)
+    perform_enqueued_jobs
+    assert_empty blocked_jobs.reload
```
I'd move this assertion to before `perform_enqueued_jobs`, because performing jobs will unblock the jobs. To verify this, check if the test passes if you remove this line: 
```ruby
 blocked_jobs.each(&:dispatch)
```
and if it does pass, it'd mean you're not quite testing the dispatching feature here.

rosa commented on test/active_job/queue_adapters/solid_queue_test.rb:
```diff
@@ -3,6 +3,8 @@
 class ActiveJob::QueueAdapters::SolidQueueTest < ActiveSupport::TestCase
   include ActiveJob::QueueAdapters::AdapterTesting
 
+  include DispatchJobs
+
```
nit: remove this empty line:
```suggestion
  include ActiveJob::QueueAdapters::AdapterTesting
  include DispatchJobs

```

rosa commented on test/dummy/db/seeds.rb:
```diff
@@ -9,21 +9,23 @@ def clean_database
 
 class JobsLoader
-  attr_reader :application, :server, :failed_jobs_count, :regular_jobs_count, :finished_jobs_count
+  attr_reader :application, :server, :failed_jobs_count, :regular_jobs_count, :finished_jobs_count, :blocked_jobs_count
```
Not directly related to your changes, but there's a method here named `dispatch_jobs`. It should be renamed to avoid confusion with the other meaning for dispatching that this PR is introducing. I suggest `perform_jobs`.

rosa commented on test/dummy/db/seeds.rb:
```diff
@@ -9,21 +9,23 @@ def clean_database
       load_finished_jobs
       load_failed_jobs
       load_regular_jobs
+      load_blocked_jobs if server.name == "solid_queue"
```
I'd rely here on the adapter's exposed supported statuses. That's it, `if server.queue_adapter.supported_statuses.include?(:blocked)` or something like that.

lexcao commented:
Hi @rosa 
Thanks for your kindly and helpful comments. I learned a lot!
I have fixed the comments.

And there still needs one help on extending `SolidQueue::Job`,
could you give me a suggestion on better do that, here is what I do for now.
```
def find_job(active_job_id)
  if job = SolidQueue::Job.find_by(active_job_id: active_job_id)
    job.extend SolidQueue::JobExt if matches_relation_filters?(job)
  end
end
```

lexcao commented:
Hey, I found there is an issue(#57) about the **dispatching** the **scheduled** job, which is closely related to this one(**blocked** job).
Should I make `dispatch` support that or another PR?

rosa commented:
> And there still needs one help on extending SolidQueue::Job,
> could you give me a suggestion on better do that, here is what I do for now.

> 
```
def find_job(active_job_id)
   if job = SolidQueue::Job.find_by(active_job_id: active_job_id)
    job.extend SolidQueue::JobExt if matches_relation_filters?(job)
  end
end
```

Hmm... could you just do this in the adapter extension? That is, why do you need to have this in `SolidQueue::Job`? If we need anything there, I'd add it directly to [Solid Queue](https://github.com/basecamp/solid_queue) instead.

lexcao commented:
OK, I just moved to the adapter.
It is not a required extension, just for the consistence of method.
```
dispatch_immediately find_solid_queue_job!(job.job_id, jobs_relation)
// or
find_solid_queue_job!(job.job_id, jobs_relation).dispatch_immediately
```

The latter keeps the consistence like this
```
find_solid_queue_job!(job.job_id, jobs_relation).retry
```

WDYT?

And the updated PR is ready for review.
Thank you!

rosa commented on lib/active_job/failed.rb:
```diff
@@ -4,8 +4,4 @@ module ActiveJob::Failed
-
-  def failed?
-    last_execution_error.present?
-  end
```
I think this might potentially cause an infinite recursion problem with this method: 
https://github.com/basecamp/mission_control-jobs/blob/6d7315b159a33e8b2c4261f322841c4324d41729/lib/active_job/executing.rb#L28-L32

Together with your change a bit below to do 
```ruby
 ActiveJob::JobsRelation::STATUSES.each do |status|
    define_method "#{status}?" do
      self.status == status
    end
  end
```
if `@status` is not defined, as we'd call `failed?`, that will try to call `self.status`, which will call `failed?` again and so on. 

This was done to support Resque and Solid Queue together. If we can guarantee that `@status` will always be defined, then we should just remove the `status` method and rely only on the attribute, which should change from `attr_writer` to `attr_accessor`. I think we can guarantee that `@status` will be set now 🤔  Initially when I modified several parts of Mission Control to work with jobs in different statuses other than `failed` and `pending`, and `failed` jobs that could be in a queue, this wasn't the case, but I think I covered all the places where we instantiate `JobProxy` and there's no other way to access a job within Mission Control.

lexcao replied:
Got it, that's really helpful.
Thank you!

rosa commented on lib/active_job/job_proxy.rb:
```diff
@@ -23,4 +23,10 @@ def serialize
+    define_method "#{status}?" do
+      self.status == status
+    end
+  end
```
See comment https://github.com/basecamp/mission_control-jobs/pull/74/files#r1505001794. If we keep this, the `status` method in  `ActiveJob::Base` (via `ActiveJob::Executing`) should go.

lexcao replied:
Yes, we need keep this to support `blocked?` method and etc.

lexcao commented:
Hi @rosa 
I have updated the last comment and fixed the conflict.
I appreciate your assistance and I am pleased to have made my first contribution to a Ruby open source project.
